src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp

```diff
@@ -0,0 +1,1420 @@
+  GEOS_HOST_DEVICE
+  void computeFlux( localIndex const iconn,
+                    StackVariables & stack,
+                    FUNC && kernelOp = NoOpFunc{} ) const
+  {
+    // first, compute the transmissibilities at this face                                             // get k and dk/dP from global arrays
+    // and place in stack
+    m_stencilWrapper.computeWeights( iconn,
+                                     m_permeability,
+                                     m_dPerm_dPres,
+                                     stack.transmissibility,
+                                     stack.dTrans_dPres );
+
+    localIndex k[2];
+    localIndex connectionIndex = 0;
+
+    for( k[0] = 0; k[0] < stack.numFluxElems; ++k[0] )
+    {
+      for( k[1] = k[0] + 1; k[1] < stack.numFluxElems; ++k[1] )
+      {
+        // clear working arrays
+        real64 densMean[numEqn]{};
+        real64 dDensMean_dP[numEqn][2]{};
+
+        real64 presGrad[numEqn]{};
+        real64 dPresGrad_dP[numEqn][2]{};
+
+        real64 gravHead[numEqn]{};
+        real64 dGravHead_dP[numEqn][2]{};
+
+        real64 capGrad[numEqn]{};
+        real64 dCapGrad_dP[numEqn][2]{};
+        real64 dCapGrad_dS[numEqn][2]{};
+
+        real64 fluxVal[numEqn]{};
+        real64 dFlux_dP[numEqn][2]{};
+        real64 dFlux_dS[numEqn][2]{};
+
+        real64 mobility[numEqn]{};
+        real64 dMob_dP[numEqn][2]{};
+        real64 dMob_dS[numEqn][2]{};
+
+        real64 const trans[2] = { stack.transmissibility[connectionIndex][0], stack.transmissibility[connectionIndex][1] };
+        real64 const dTrans_dP[2] = { stack.dTrans_dPres[connectionIndex][0], stack.dTrans_dPres[connectionIndex][1] };
+
+        // cell indices
+        localIndex const seri[2]  = {m_seri( iconn, k[0] ), m_seri( iconn, k[1] )};
+        localIndex const sesri[2] = {m_sesri( iconn, k[0] ), m_sesri( iconn, k[1] )};
+        localIndex const sei[2]   = {m_sei( iconn, k[0] ), m_sei( iconn, k[1] )};
+
+        // loop over phases
+        for( integer ip = 0; ip < m_numPhases; ++ip )
+        {
+          // calculate quantities on primary connected cells
+          for( integer ke = 0; ke < 2; ++ke )
+          {
+            // density
+            real64 const density  = m_dens[seri[ke]][sesri[ke]][sei[ke]][0][ip];         // r = rho1 || rho2
+            real64 const dDens_dP = m_dDens_dPres[seri[ke]][sesri[ke]][sei[ke]][0][ip][Deriv::dP];  // dr/dP = dr1/dP1 || dr2/dP
+
+            // average density and derivatives
+            densMean[ip] += 0.5 * density;          // rho = (rho1 + rho2) / 2
+            dDensMean_dP[ip][ke] = 0.5 * dDens_dP;  // drho/dP = { (dr1/dP1)/2 , (dr2/dP2)/2 }
+          }
+
+          //***** calculation of flux *****
+
+          // compute potential difference
+          real64 potScale = 0.0;
+          real64 dPresGrad_dTrans = 0.0;
+          real64 dGravHead_dTrans = 0.0;
+          real64 dCapGrad_dTrans = 0.0;
+          int signPotDiff[2] = {1, -1};
```

Comment:
i don't think you need this sign thing

Reply:
Keeping it for now. We can think of a more elegant implementation in the future.

Reply:
If we understand correctly, it's because we need a pressure difference instead of a single pressure value:

Here presGrad = T * (P1-P2)
When we take the derivative with respect to the two pressures we will need: dPresGrad_dP = {dT/dP1 * (P1-P2) + T, dT/dP1 * (P1-P2) - T}. At the last line of code shown dPresGrad_dP = {T, -T} already.
So the second term on the right-hand side needs to be a (P1-P2), instead of just a single pressure. We used the signPotDiff variable to assist taking the difference of the 2 pressures.

Reply:
i am still confused
the code is more like
`presGrad = T1*P1 + T2*P2`
right?
the rest should follow the same notation/logic
maybe @CusiniM or @ryar9534 can have a look and check if i am wrong

Reply:
Someone else can check our understanding, but it's because in
 `presGrad[ip] += trans[ke] * pressure;    // DPv = T (P1 - P2)` 
we are using the transmissibilities that come from the `computeWeights (CellElementStencilTPFA.hpp)` function like this:
`weight[0][ke] = m_transMultiplier[iconn] * value * (ke == 0 ? 1 : -1);`
So T2 = -T1